cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,983 @@
+            op.expected_type = OBJECT;
+          }
+          return op;
+        }
+      } break;
+
+      // 3 ways this can be used
+      // indices:   [0]
+      // name:      ['book']
+      // wildcard:  [*]
+      case '[': {
+        path_operator op;
+        json_string term{"]", 1};
+        bool const is_string = *pos == '\'' ? true : false;
+        if (parse_path_name(op.name, term)) {
+          pos++;
+          if (op.name.len == 1 && op.name.str[0] == '*') {
+            op.type          = path_operator_type::CHILD_WILDCARD;
+            op.expected_type = NONE;
+          } else {
+            if (is_string) {
+              op.type          = path_operator_type::CHILD;
+              op.expected_type = OBJECT;
+            } else {
+              op.type = path_operator_type::CHILD_INDEX;
+              op.index =
+                cudf::io::parse_numeric<int>(op.name.str, op.name.str + op.name.len, json_opts, -1);
+              CUDF_EXPECTS(op.index >= 0, "Invalid numeric index specified in JSONPath");
+              op.expected_type = ARRAY;
+            }
+          }
+          return op;
+        }
+      } break;
+
+      // wildcard operator
+      case '*': {
+        pos++;
+        return path_operator{path_operator_type::CHILD_WILDCARD};
+      } break;
+
+      default: CUDF_FAIL("Unrecognized JSONPath operator"); break;
+    }
+    return {path_operator_type::ERROR};
+  }
+
+ private:
+  cudf::io::parse_options_view json_opts{',', '\n', '\"', '.'};
+
+  bool parse_path_name(json_string& name, json_string const& terminators)
+  {
+    switch (*pos) {
+      case '*':
+        name.str = pos;
+        name.len = 1;
+        pos++;
+        break;
+
+      case '\'':
+        if (parse_string(name, false, '\'') != parse_result::SUCCESS) { return false; }
+        break;
+
+      default: {
+        size_t const chars_left = input_len - (pos - input);
+        char const* end         = std::find_first_of(
+          pos, pos + chars_left, terminators.str, terminators.str + terminators.len);
+        if (end) {
+          name.str = pos;
+          name.len = end - pos;
+          pos      = end;
+        } else {
+          name.str = pos;
+          name.len = chars_left;
+          pos      = input + input_len;
+        }
+        break;
+      }
+    }
+
+    // an empty name is not valid
+    CUDF_EXPECTS(name.len > 0, "Invalid empty name in JSONpath query string");
+
+    return true;
+  }
+};
+
+/**
+ * @brief Preprocess the incoming JSONPath string on the host to generate a
+ * command buffer for use by the GPU.
+ *
+ * @param json_path The incoming json path
+ * @param stream Cuda stream to perform any gpu actions on
+ * @returns A tuple containing the command buffer, the maximum stack depth required and whether or
+ * not the command buffer is empty.
+ */
+std::tuple<rmm::device_uvector<path_operator>, int, bool> build_command_buffer(
+  cudf::string_scalar const& json_path, rmm::cuda_stream_view stream)
```

Comment:
Another case where a `bool` parameter can be more intuitively replaced with an `optional`.